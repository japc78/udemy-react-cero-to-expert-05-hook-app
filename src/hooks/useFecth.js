import { useEffect, useState } from 'react'

export const useFecth = ( url ) => {
	const [state, setState] = useState({ data: null, loading: true, error: null})

	useEffect(() => {

		//Para que salga siempre loading al leer un nuevo elemento
		setState ( {
			...state,
			loading: true
		})

		fetch(url)
			.then( resp => resp.json() )
			.then( data => {
				setState({
					loading: false,
					error: null,
					data
				})
			})

	}, [url])

	return state
}
