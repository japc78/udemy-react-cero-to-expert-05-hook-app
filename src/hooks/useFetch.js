import { useEffect, useRef, useState } from 'react'

export const useFetch = ( url ) => {
	const isMounted = useRef(true);

	const [state, setState] = useState({ data: null, loading: true, error: null});


	// Creamos un useEffect para poder controlar si el componente esta montado o no, [] en el segundo atributo indica que solo se ejecutara el use effects cuando se desmonte el componente y en ese caso cambiamos el valor del useRef isMounted a false.
	useEffect(() => {
		return () => {
			isMounted.current = false;
		}
	}, []);


	useEffect(() => {

		//Para que salga siempre loading al leer un nuevo elemento
		setState ( {
			...state,
			loading: true
		})

		fetch(url)
			.then( resp => resp.json() )
			.then( data => {

				//Si es verdadero monta el componente.
				if (isMounted.current) {
					setState({
						loading: false,
						error: null,
						data
					})
				}
			})
			.catch(()=> {
				setState({
					data: null,
					loading: false,
					error: 'No se pudo cargar la info'
				})
			});

	}, [url]);

	return state
}
