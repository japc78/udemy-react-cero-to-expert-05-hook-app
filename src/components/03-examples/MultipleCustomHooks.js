import React from 'react'
import { useCounter } from '../../hooks/useCounter'
import { useFetch } from '../../hooks/useFetch'
import './styles.css'
export const MultipleCustomHooks = () => {

	const { counter, increment } = useCounter(1);

	// const state = useFecth('https://www.breakingbadapi.com/api/quotes/1');
	const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);

	//doble negacion !! esplicacion: !data, sino tiene valor, !!data == data, es decir si tiene valor.
	// https://developer.mozilla.org/es/docs/Web/JavaScript/Referencia/Operadores/Operadores_l%C3%B3gicos
	const { author, quote } = !!data && data[0];

	// console.log(loading);

	return (
		<>
			<h1>BreakingBad Quotes</h1>
			<hr/>

			{
				loading
				?
				(<div className="alert alert-info">loading...</div>)
				:
				(
					<blockquote className="blockquote text-right">
						<p className="mb-0">{quote}</p>
						<footer className="blockquote-footer">{author}</footer>
					</blockquote>
				)
			}

			<button className="btn btn-primary" onClick={ increment }>Siguiente frase</button>
		</>
	)
}
