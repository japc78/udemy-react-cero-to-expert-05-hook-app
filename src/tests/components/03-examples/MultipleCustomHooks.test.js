import '@testing-library/jest-dom'
import { shallow } from 'enzyme'
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks'
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';


jest.mock('../../../hooks/useFetch')
jest.mock('../../../hooks/useCounter')


describe('Test on <MultipleCustomHooks/>', () => {

	useCounter.mockReturnValue({
		counter: 1,
		increment: () => {}
	});


	test('should show it correctly', () => {
		useFetch.mockReturnValue({
			data: null,
			loading: true,
			error: null
		});

		// useCounter.mockReturnValue({
		// 	counter: 1,
		// 	increment: () => {}
		// });

		const wrapper = shallow(<MultipleCustomHooks/>);
		expect( wrapper ).toMatchSnapshot();
	});

	test('should show the information', () => {
		useFetch.mockReturnValue({
			data: [{
				author: 'Fernando',
				quote: 'Hola Mundo'
			}],
			loading: false,
			error: null
		});

		// useCounter.mockReturnValue({
		// 	counter: 1,
		// 	increment: () => {}
		// });

		const wrapper = shallow(<MultipleCustomHooks/>);
		// console.log(wrapper.html());

		expect(wrapper.find('.alert').exists()).toBe(false);
		expect(wrapper.find('.mb-0').text().trim()).toBe('Hola Mundo');
		expect(wrapper.find('footer').text().trim()).toBe('Fernando');
	})
})
