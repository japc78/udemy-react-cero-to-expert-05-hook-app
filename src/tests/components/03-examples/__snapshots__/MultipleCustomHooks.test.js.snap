// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <MultipleCustomHooks/> should show it correctly 1`] = `
<Fragment>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info"
  >
    loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente frase
  </button>
</Fragment>
`;
